File: solutions/python/zipper/1/zipper.py
class Zipper:
    def __init__(self, tree, crumbs=None):
        self._tree = tree
        self._crumbs = crumbs or []

    @staticmethod
    def from_tree(tree):
        return Zipper(tree)

    def value(self):
        return self._tree["value"]

    def set_value(self, value):
        return Zipper(
            {"value": value, "left": self._tree["left"], "right": self._tree["right"]},
            self._crumbs,
        )

    def left(self):
        if self._tree["left"] is None:
            return None
        return Zipper(
            self._tree["left"],
            self._crumbs + [("left", self._tree["value"], self._tree["right"])],
        )

    def set_left(self, tree):
        return Zipper(
            {"value": self._tree["value"], "left": tree, "right": self._tree["right"]},
            self._crumbs,
        )

    def right(self):
        if self._tree["right"] is None:
            return None
        return Zipper(
            self._tree["right"],
            self._crumbs + [("right", self._tree["value"], self._tree["left"])],
        )

    def set_right(self, tree):
        return Zipper(
            {"value": self._tree["value"], "left": self._tree["left"], "right": tree},
            self._crumbs,
        )

    def up(self):
        if not self._crumbs:
            return None
        direction, val, sibling = self._crumbs[-1]
        rest = self._crumbs[:-1]
        if direction == "left":
            return Zipper(
                {"value": val, "left": self._tree, "right": sibling}, rest
            )
        return Zipper(
            {"value": val, "left": sibling, "right": self._tree}, rest
        )

    def to_tree(self):
        z = self
        while True:
            u = z.up()
            if u is None:
                break
            z = u
        return z._tree

    def __eq__(self, other):
        return self.to_tree() == other.to_tree()

File: solutions/ruby/zipper/1/zipper.rb
Node = Struct.new(:value, :left, :right)

class Zipper
  def self.from_tree(tree)
    new(tree)
  end

  def initialize(tree, crumbs = [], left_tree = nil, right_tree = nil)
    @tree = tree
    @crumbs = crumbs
  end

  def value
    @tree.value
  end

  def left
    return nil unless @tree.left
    Zipper.new(@tree.left, @crumbs + [[:left, @tree.value, @tree.right]])
  end

  def right
    return nil unless @tree.right
    Zipper.new(@tree.right, @crumbs + [[:right, @tree.value, @tree.left]])
  end

  def up
    return nil if @crumbs.empty?
    dir, val, sibling = @crumbs.last
    if dir == :left
      Zipper.new(Node.new(val, @tree, sibling), @crumbs[0...-1])
    else
      Zipper.new(Node.new(val, sibling, @tree), @crumbs[0...-1])
    end
  end

  def set_value(value)
    Zipper.new(Node.new(value, @tree.left, @tree.right), @crumbs)
  end

  def set_left(tree)
    Zipper.new(Node.new(@tree.value, tree, @tree.right), @crumbs)
  end

  def set_right(tree)
    Zipper.new(Node.new(@tree.value, @tree.left, tree), @crumbs)
  end

  def to_tree
    z = self
    loop do
      u = z.up
      break unless u
      z = u
    end
    z.instance_variable_get(:@tree)
  end

  def ==(other)
    to_tree == other.to_tree
  end
end
